arangod/RocksDBEngine/RocksDBMetaCollection.cpp

```diff
@@ -450,114 +458,304 @@ std::unique_ptr<containers::RevisionTree> RocksDBMetaCollection::computeRevision
+    // with its commit marker at a sequence number between
+    // `_revisionTreeSerializedSeq` and `seq` (i.e. its
+    // `postCommitSeq`). Then it must have before got a blocker for its
+    // `beginSeq` number in Step 2 above. So either, we still see the blocker,
+    // or the corresponding operations have already been added to the buffers,
+    // or they have already been applied to the tree, in which case
+    // `_revisionTreeApplied` would be larger than `seq`. q.e.d.
+    LOG_TOPIC("32d45", TRACE, Logger::ENGINES)
+        << "adjusting sequence number for " << _logicalCollection.name() 
+        << " from " << _revisionTreeSerializedSeq << " to " << seq;
+    
+    _revisionTreeSerializedSeq = seq;
+    _revisionTreeApplied = seq;
+    // This allows to uphold the invariants that
+    // _revisionTreeCreationSeq <= _revisionTreeSerializedSeq and
+    // _revisionTreeSerializedSeq <= _revisionTreeApplied.
   }
+  
+  return _revisionTreeSerializedSeq;
+}
 
-  if (!_revisionInsertBuffers.empty() && _revisionInsertBuffers.begin()->first - 1 < seq) {
-    seq = _revisionInsertBuffers.begin()->first - 1;
-  }
+rocksdb::SequenceNumber RocksDBMetaCollection::serializeRevisionTree(
+    std::string& output, rocksdb::SequenceNumber commitSeq, bool force,
+    std::unique_lock<std::mutex> const& lock) {
+  TRI_ASSERT(lock.owns_lock());
+  TRI_ASSERT(_logicalCollection.useSyncByRevision());
 
-  if (!_revisionRemovalBuffers.empty() && _revisionRemovalBuffers.begin()->first - 1 <= seq) {
-    seq = _revisionRemovalBuffers.begin()->first - 1;
+  if (!_revisionTree && !haveBufferedOperations(lock)) {  // empty collection
+    return commitSeq;
   }
-
-  // limit to before the last thing we applied, since we haven't persisted it
-  rocksdb::SequenceNumber applied = _revisionTreeApplied.load();
-  if (applied > _revisionTreeSerializedSeq && applied - 1 < seq) {
-    seq = applied - 1;
+  if (!_revisionTreeCanBeSerialized) {
+    // previously, a revision tree was found to be bad and is now being repaired, 
+    // in the meantime we return the number when it was last persisted to avoid 
+    // lastSync jumping down again later
+    return _revisionTreeSerializedSeq;
   }
 
-  // now actually advance it if we can
-  if (seq > _revisionTreeSerializedSeq) {
-    _revisionTreeSerializedSeq = seq;
-  }
+  applyUpdates(commitSeq, lock);  // always apply updates...
 
-  return _revisionTreeSerializedSeq;
-}
+  // applyUpdates will make sure we will have a valid tree
+  TRI_ASSERT(_revisionTree != nullptr);
+  TRI_ASSERT(_revisionTree->depth() == revisionTreeDepth);
 
-rocksdb::SequenceNumber RocksDBMetaCollection::serializeRevisionTree(
-    std::string& output, rocksdb::SequenceNumber commitSeq, bool force) {
-  std::unique_lock<std::mutex> guard(_revisionTreeLock);
+  bool neverDone = _revisionTreeSerializedSeq <= _revisionTreeCreationSeq;
+  bool coinFlip = RandomGenerator::interval(static_cast<uint32_t>(5)) == 0;
```

Comment:
We want to prevent to write too often if the collection has a lot of write traffic.

Reply:
No, the coin flip always existed and it also makes sense.
We do not want to write data for unchanged collections every time to reduce background write activity. By doing the coin flip, we only write out the data every `coinFlip`th time. It is random so that we do not have the pattern `true, true, true, true, false` for all collections at the same time, but randomly distribute the load so it is evened out.